lib/saluki-components/src/sources/checks/python_scheduler.rs

```diff
@@ -0,0 +1,486 @@
+                        can_run,
+                        name = module_name,
+                    );
+                    if can_run {
+                        return true;
+                    }
+                }
+                Err(e) => {
+                    error!("Python panic during module check for {}: {:?}", module_name, e);
+                    // Continue trying other modules
+                }
+            }
+        }
+
+        false
+    }
+
+    // This function does 3 things
+    // 1. Registers the check source code and gets a handle
+    // 2. Starts a local task for each instance that
+    //    queues a run of the check every min_collection_interval_ms
+    // 3. Stores the handles in the running hashmap
+    fn run_check(&mut self, check: &RunnableCheckRequest) -> Result<(), GenericError> {
+        let check_handle = self.register_check(check)?;
+        let running_entry = self
+            .running
+            .entry(check.check_request.source.clone())
+            .or_insert((check_handle.clone(), Vec::new()));
+
+        info!(
+            "Running check {name} with {num_instances} instances",
+            name = check.check_request.name,
+            num_instances = check.check_request.instances.len()
+        );
+        for (idx, instance) in check.check_request.instances.iter().enumerate() {
+            let instance = instance.clone();
+            let init_config = check.check_request.init_config.clone();
+
+            let check_handle = check_handle.clone();
+            trace!("Spawning task for check instance {idx}");
+            self.tlm.check_instances_started.increment(1);
+            let handle = tokio::task::spawn(async move {
+                let mut interval =
+                    tokio::time::interval(Duration::from_millis(instance.min_collection_interval_ms().into()));
+                loop {
+                    interval.tick().await;
+                    // run check
+                    info!("Running check instance {idx}");
+                    pyo3::Python::with_gil(|py| {
+                        let instance_as_pydict = instance.to_pydict(&py);
+
+                        let instance_list = PyList::new_bound(py, &[instance_as_pydict]);
+                        let kwargs = PyDict::new_bound(py);
+                        kwargs
+                            .set_item("name", "placeholder_check_name")
+                            .expect("Could not set name");
+                        kwargs
+                            .set_item("init_config", init_config.to_pydict(&py))
+                            .expect("could not set init_config");
+                        kwargs
+                            .set_item("instances", instance_list)
+                            .expect("could not set instance list");
+
+                        // TODO should this be moved up outside of the loop?
+                        // I kind of think so, this makes a new check instance every tick
```

Comment:
Worth highlighting this 😅 , this is the kind of thing I mentioned that @ofek would have a much better grasp on